# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1710007597 -3600
# Parent  916b920af38efc3b75a348699d05aecefaeb5b05
Rework MailNotificationService.jsm to improve Windows systray tooltip: Fix logic error.

diff --git a/mailnews/base/src/MailNotificationService.jsm b/mailnews/base/src/MailNotificationService.jsm
--- a/mailnews/base/src/MailNotificationService.jsm
+++ b/mailnews/base/src/MailNotificationService.jsm
@@ -218,12 +218,24 @@ class NewMailNotificationService {
       );
       return;
     }
 
+    let newFolderCount = Math.max(folderCount + newCount, 0);
     this.#log.debug(
-      `_updateNewCount: ${folder.URI}, setting ${folderCount} + ${newCount}`
+      `_updateNewCount: ${folder.URI}, setting ${folderCount} + ${newCount} = ${newFolderCount}`
     );
-    this.folderCountMap.set(folder.URI, Math.max(folderCount + newCount, 0));
+    if (newFolderCount > 0) {
+      this.folderCountMap.set(folder.URI, newFolderCount);
+    } else {
+      // Likely this is the second call on a compensation, so make sure the
+      // maps don't hang around. If the maps hang around, the condition above:
+      // if (newCount < 0 && folderCountMapValue === undefined) { ...
+      // won't be fulfilled again.
+      this.#log.debug(`_updateNewCount: ${folder.URI}: Deleting count map`);
+      this.folderCountMap.delete(folder.URI);
+      // Don't delete the folder map or count may not update in the UI below.
+      // this.folderNewMap.delete(folder.URI);
+    }
     this.#newCount = Math.max(this.#newCount + newCount, 0);
     this.#log.debug(
       `_updateNewCount: ${folder.URI} new mail count: ${this.#newCount}`
     );
